packages/realm/src/RealmListeners.ts

```diff
@@ -0,0 +1,92 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2022 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import { Realm } from "./internal";
+
+export type RealmEventName = "change" | "schema" | "beforenotify";
+enum RealmEvent {
+  Change = "change",
+  Schema = "schema",
+  BeforeNotify = "beforenotify",
+}
+export type RealmListenerCallback = (r: Realm, name: RealmEventName, schema?: Realm.ObjectSchema[]) => void;
+
+// Temporary functions to work between event names and corresponding enums.
```

Comment:
this is what I mean by it feels a little unnecessary to do all this just to store the name as an enum but if we do not want to change user-facing API this is useful for future change.

Reply:
You could do something a bit more reflective like this: https://www.typescriptlang.org/play?#code/KYOwrgtgBASsCGAbCBRAbqALgOXhYUA3gFBRQDCAFvCAOYEC8UARAMbV3DMA0pUAyu2AR4UJswDOQkTz4AhYADMA9gCdg2ZZgCWigJ5iWAIyVrQW3XtkBfYsUVgQrHcpBRtE9Flz4AFCDxgAC4oCUxVbToAShD4CQlgVUwJKAD8dxS4JFQMEBxAoj5dKF8AQgB5IwArYGcAOjQkMGAJXyzkLzyfYCi6yNZEMAATFv9AqKjCsjJMSlVlAHdU4CWUVXnVXwADAFUQYAAPAAdazGAhqGBczFSCgHIAEkI04Gs7raiAbj5bW3tHZzaVxQFTKXxXLAhMIRaJTDKdfJ+CF5L58IzwTbIzCov4OJwuNzozHXELtHLecZw1iuCTKRDAOqIZS0XzMIlQVhIekXBbaWY8S7XHF2UGs9g0ejMVGi5gLeZ0KXfIA

Reply:
I think this assertion already ends up being indirectly being done by the Real as it ends up looking through all possible values before running this function for the specific one. If you mean for constructing these listener objects, they get initialised in the Realm with constant string, so Typescript ends up asserting types anyways.